crates/matrix-sdk-ui/src/sync_service/mod.rs

```diff
@@ -72,118 +104,295 @@ impl SyncService {
-                        }
+            let (is_error, has_expired) = loop {
+                let res = room_list_stream.next().await;
+                match res {
+                    Some(Ok(())) => {
+                        // Carry on.
                     }
-                }
-            } else {
-                while let Some(res) = room_list_stream.next().await {
-                    if let Err(err) = res {
-                        tracing::error!("Error while processing app (room list) state: {err:#}");
-                        state_observer.set(SyncServiceState::Error);
-                        break;
+                    Some(Err(err)) => {
+                        // If the room list error was an expired session, also expire the
+                        // encryption sync.
+                        let has_expired = if let room_list_service::Error::SlidingSync(err) = &err {
+                            err.client_api_error_kind()
+                                == Some(&ruma::api::client::error::ErrorKind::UnknownPos)
+                        } else {
+                            false
+                        };
+                        error!("Error while processing room list in sync service: {err:#}");
+                        break (true, has_expired);
+                    }
+                    None => {
+                        // The stream has ended.
+                        break (false, false);
                     }
                 }
+            };
+
+            if let Err(err) = sender
+                .send(TerminationReport {
+                    is_error,
+                    has_expired,
+                    origin: TerminationOrigin::RoomList,
+                })
+                .await
+            {
+                error!("Error while sending termination report: {err:#}");
+            }
+        }
+    }
+
+    /// Start (or restart) the underlying sliding syncs.
+    ///
+    /// This can be called multiple times safely:
+    /// - if the stream is still properly running, it won't be restarted.
+    /// - if the stream has been aborted before, it will be properly cleaned up
+    ///   and restarted.
+    pub async fn start(&self) {
+        let _guard = self.modifying_state.lock().await;
+
+        // Only (re)start the tasks if any was stopped.
+        if matches!(self.state.get(), SyncServiceState::Running) {
+            // It was already true, so we can skip the restart.
+            return;
+        }
 
-                state_observer.set(SyncServiceState::Terminated);
+        trace!("starting sync service");
+
+        let (sender, receiver) = tokio::sync::mpsc::channel(16);
+
+        // First, take care of the room list.
+        {
+            let mut room_list_task_guard = self.room_list_task.lock().unwrap();
+            if let Some(task) = room_list_task_guard.take() {
+                warn!("active room list service task when start()ing sync service");
+                task.abort();
```

Comment:
Why aborting here? In case we call `start` twice without `stop`ping it?

Reply:
It's really me being super paranoid that things end up in a weird state, and I prefer to have self-healing code, but this really should NEVER happen (hence the "warn"). Thoughts how to handle that better?